sources/modules/program-design/iterators.md

```diff
@@ -0,0 +1,138 @@
+_so please don't remove these descriptions when editing the topic._
+
+### Overview
+
+_Provides a short natural language abstract of the module’s contents._
+_Specifies the different levels of teaching._
+
+------------------------------------------------------------------------
+Level             Objective
+----------------- ------------------------------------------------------
+Foundational      Using iterators
+
+Main              Writing iterators
+
+Advanced          ---
+
+------------------------------------------------------------------------
+
+### Motivation
+
+_Why is this important?_
+_Why do we want to learn/teach this topic?_
+
+Iterators provide a generic interface between algorithms and containers.
+The container defines the iteration over its data.
+Therefore, the algorithm does neither need to know about the iteration nor the concrete data types that are stored in the container.
+This way, algorithms can be implemented in a generic way without detailed knowledge of the containers they operate on, and the containers can be written without predetermining what algorithms should be available for it.
+
+### Topic introduction
+
+_Very brief introduction to the topic._
+
+### Foundational: Using iterators {#it-found}
+
+#### Background/Required Knowledge
+
+A student:
+* containers [[Program Design: Containers - Foundational]][1]
+* for loops
+* [Maybe] [[Object Model: Value Semantics - Foundational]][2][[Object Model: Reference Semantics - Foundational]][3] TODO: revisited after ref/value semantics 
+
+#### Student outcomes
+
+_A list of things "a student should be able to" after the curriculum._
+_The next word should be an action word and testable in an exam._
+_Max 5 items._
+
+A student should be able to:
+
+1. write a for loop using iterators (not just range-based for).
+2. use the canonical iterator operations (begin/end, dereference, etc.).
+3. use the appropriate operations with regard to the iterators category.
+4. show examples where iterators are invalidated by operations on a container.
+5. demonstrate how iterators provide an abstraction between the operation and the underlying data structure.
+
+#### Caveats
+
+_This section mentions subtle points to understand, like anything resulting in
+implementation-defined, unspecified, or undefined behavior._
+
+* Dangling iterators (i.e., having an iterator to a no longer valid container)
+* `end()` marks the position after the last element and, by that, is not dereferencable
+* When comparing iterators, both iterators need to belong to the same underlying data structure
+* Keep in mind that iterators do not guarantee that the underlying memory is contiguous.
+
+#### Points to cover
+
+_This section lists important details for each point._
+
+* Using standard iterators with iterator-based for loops
+    * Difference between value/reference semantics based for loops
+    * Explain connection to for-each
+* Basic iterator nomenclature and functionality
+    * begin/end
```

Comment:
Explain that begin/end are methods for the container class, but advance/dereference are methods for the iterator class, which is a ?subclass? of the container class.